### PR TITLE
Added locality attributes for Typeclasses Opaque and Transparent

### LINKED ISSUE #14513: Support "Local" and "Global" annotations for "Typeclasses Opaque"
#### Description of the problem

Currently, it is very easy to accidentally put `Typeclasses Opaque` into a `Section`. This means its effect will be lost when the section closes, which often is not desired. This is hard to notice or debug.

It would be great if `Typeclasses Opaque` supported explicit `Local` and `Global` just like the underlying `Hint Opaque`.

#### Coq Version

8.13.2

### PR DESCRIPTION
**Pull request by Alizter**
We extend the grammar of `Typeclasses Opaque` and `Typeclasses Transparent` to include locality attributes, and then pass these attributes to the underlying `Hint Opaque` command. Like the underlying command, we deprecate usage without an attribute.

Fixes / closes #14513

- [x] Added / updated **test-suite**.

- [x] Added **changelog**.
- [x] Added / updated **documentation**.

Should I document the fact that `Typeclasses Opaque` will throw the Hint warning?

Overlays:
- https://github.com/mattam82/Coq-Equations/pull/439

### CODE REVIEW

File: vernac/classes.ml [hunk 1]
```diff
@@ -28,8 +28,8 @@ module RelDecl = Context.Rel.Declaration
 module NamedDecl = Context.Named.Declaration
 (*i*)
 
-let set_typeclass_transparency c local b =
-  let locality = if local then Goptions.OptLocal else Goptions.OptGlobal in
+let set_typeclass_transparency ?(locality=Goptions.OptLocal) c b =
```
Comment (SkySkimmer) [inline vernac/classes.ml:31]:
```suggestion
let set_typeclass_transparency ~locality c b =
```
no need for this to be implicit

Comment (Alizter) [reply vernac/classes.ml:31]: Why not?

Comment (SkySkimmer) [reply vernac/classes.ml:31]: Implicit arguments make it harder to tell what's going on so should be avoided unless there's a benefit

File: plugins/ltac/g_class.mlg [hunk 2]
```diff
@@ -22,22 +22,27 @@ DECLARE PLUGIN "ltac_plugin"
 
 {
 
-let set_transparency cl b =
+let set_transparency locality cl b =
+  let mod_locality = match locality with
+    |  Goptions.OptDefault -> if Locality.make_section_locality None then Goptions.OptGlobal else Goptions.OptLocal
```
Comment (SkySkimmer) [inline plugins/ltac/g_class.mlg:27]:
```suggestion
    |  Goptions.OptDefault -> if not (Global.sections_are_opened ()) then Goptions.OptGlobal else Goptions.OptLocal
```

File: vernac/record.ml [hunk 3]
```diff
@@ -640,7 +640,8 @@ let build_class_constant ~univs ~rdata field implfs params paramimpls coers bind
   let cref = GlobRef.ConstRef cst in
   Impargs.declare_manual_implicits false cref paramimpls;
   Impargs.declare_manual_implicits false (GlobRef.ConstRef proj_cst) (List.hd implfs);
-  Classes.set_typeclass_transparency (Tacred.EvalConstRef cst) false false;
+  Classes.set_typeclass_transparency ~locality:Goptions.OptLocal
```
Comment (SkySkimmer) [inline vernac/record.ml:643]: shouldn't it be global? also the one below 
(previously false = not local)

File: plugins/ltac/g_class.mlg (hunk 2 above)
Comment (Alizter) [reply plugins/ltac/g_class.mlg:27]: Should I just swap the order instead of having a not?

File: vernac/record.ml (hunk 3 above)
Comment (Alizter) [reply vernac/record.ml:643]: OK, that has confused me

Comment (Alizter) [reply vernac/record.ml:643]: When I change it to global or export, it complains that it is not allowed in sections.

**Comment by ppedrot** [comment]
Note that in #14697 the locality checks are deferred to the low-level hint functions, so if it gets merged first this should simplify this PR quite a bit (and remove the TODOs).

**Comment by Alizter** [comment]
@ppedrot Thanks! I'll keep this as a draft for the time being.

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@Alizter why did you remove the milestone?

**Review by Zimmi48 — COMMENTED** [review]
Documentation and changelog are incomplete: it should be mentioned that calling `Typeclasses Opaque` / `Transparent` without a locality attribute is deprecated (because the default locality will change in the future).

**Review by ppedrot — APPROVED** [review]

**Review by Zimmi48 — APPROVED** [review]
I'd prefer if the changelog was more explicit about the deprecation, but that's not blocking.

**Pull request merged**